File: app-shared/Sources/CommonLibraryCore/Strategy/GitHubConfigStrategy.swift
// SPDX-FileCopyrightText: 2026 Davide De Rosa
//
// SPDX-License-Identifier: GPL-3.0

import MiniFoundation

@MainActor
public final class GitHubConfigStrategy: ConfigManagerStrategy {
    private let logger: AppLogger

    private let url: URL

    private let betaURL: URL

    private let ttl: TimeInterval

    private let isBeta: @Sendable () async -> Bool

    private let fetcher: @Sendable (URL) async throws -> Data

    private var lastUpdated: Date

    public init(
        _ logger: AppLogger,
        url: URL,
        betaURL: URL,
        ttl: TimeInterval,
        isBeta: @escaping @Sendable () async -> Bool,
        fetcher: @escaping @Sendable (URL) async throws -> Data
    ) {
        self.logger = logger
        self.url = url
        self.betaURL = betaURL
        self.ttl = ttl
        self.isBeta = isBeta
        self.fetcher = fetcher
        lastUpdated = .distantPast
    }

    public func bundle() async throws -> ABI.ConfigBundle {
        let isBeta = await isBeta()
        logger.log(.core, .debug, "Config (GitHub): beta = \(isBeta)")
        if lastUpdated > .distantPast {
            let elapsed = -lastUpdated.timeIntervalSinceNow
            let ttl = isBeta ? ttl / 10.0 : ttl
            guard elapsed >= ttl else {
                logger.log(.core, .debug, "Config (GitHub): elapsed \(elapsed) < \(ttl)")
                throw ABI.AppError.rateLimit
            }
        }
        let targetURL = isBeta ? betaURL : url
        logger.log(.core, .info, "Config (GitHub): fetching bundle from \(targetURL)")
        let data = try await fetcher(targetURL)
        let json = try JSONDecoder().decode(ABI.ConfigBundle.self, from: data)
        lastUpdated = Date()
        return json
    }
}

File: app-android/app/src/main/kotlin/com/algoritmico/passepartout/strategy/GitHubConfigStrategy.kt
// SPDX-FileCopyrightText: 2026 Davide De Rosa
//
// SPDX-License-Identifier: GPL-3.0

package com.algoritmico.passepartout.strategy

import android.os.SystemClock
import android.util.Log
import com.algoritmico.passepartout.managers.ConfigBundle
import com.algoritmico.passepartout.managers.ConfigManagerException
import com.algoritmico.passepartout.managers.ConfigManagerStrategy

class GitHubConfigStrategy(
    private val logTag: String,
    private val url: String,
    private val ttl: Double,
    private val fetcher: suspend (String) -> ByteArray
) : ConfigManagerStrategy {
    private var lastUpdatedAtMillis: Long? = null

    override suspend fun bundle(): ConfigBundle {
        val now = SystemClock.elapsedRealtime()
        lastUpdatedAtMillis?.let { lastUpdatedAtMillis ->
            val elapsed = (now - lastUpdatedAtMillis) / 1000.0
            if (elapsed < ttl) {
                Log.d(logTag, "Config (GitHub): elapsed $elapsed < $ttl")
                throw ConfigManagerException.RateLimit
            }
        }

        Log.i(logTag, "Config (GitHub): fetching bundle from $url")
        val data = fetcher(url)
        val bundle = ConfigBundle.decode(data)
        lastUpdatedAtMillis = SystemClock.elapsedRealtime()
        return bundle
    }
}
